iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -0,0 +1,199 @@
+  if (workgroupSize[2] != 1) {
+    return op->emitOpError("expected workgroup z component to be 1 for ")
+           << pipelineName << ", got " << workgroupSize[2];
+  }
+
+  // Verify shared memory usage of operands after tiling
+  // requires <= 64Kb combined space.
+  SmallVector<int64_t> firstLevelTileSizes =
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  MemRefType type =
+      matmulOp.getInputOperand(0)->get().getType().cast<mlir::MemRefType>();
+  unsigned bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+
+  // Input shape sizes: A [ M x K],  B [ K x N]
+  unsigned totalSharedMemSizeBytes =
+      (firstLevelTileSizes[0] * firstLevelTileSizes[2] +
+       firstLevelTileSizes[1] * firstLevelTileSizes[2]) *
+      bytesSize;
+
+  if (totalSharedMemSizeBytes > kSharedMemSizeBytes) {
+    return op->emitOpError("expected shared memory usage <= 64Kb for ")
+           << pipelineName << ", got " << totalSharedMemSizeBytes;
+  }
+
+  return success();
+}
+
+LogicalResult verifyGPUMatmulTensorCorePipeline(
+    Operation *op, IREE::Codegen::LoweringConfigAttr loweringConfig,
+    IREE::Codegen::TranslationInfoAttr translationInfo,
+    ArrayRef<int64_t> workgroupSize) {
+  auto pipeline =
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulTensorCore;
+  StringRef pipelineName = stringifyEnum(pipeline);
+  if (workgroupSize.empty()) {
+    return op->emitOpError("expected workgroup size for GPU pipelines");
+  }
+
+  ArrayRef<int64_t> lhsShape;
+  ArrayRef<int64_t> rhsShape;
+  Type inputType;
+  SmallVector<int64_t> firstLevelTileSizes;
+  if (linalg::MatmulOp matmulOp = dyn_cast<linalg::MatmulOp>(op)) {
+    inputType = matmulOp.getInputOperand(0)->get().getType();
+    lhsShape = getUntiledShape(matmulOp.getInputOperand(0)->get());
+    rhsShape = getUntiledShape(matmulOp.getInputOperand(1)->get());
+    firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  } else if (linalg::BatchMatmulOp batchMatmulOp =
+                 dyn_cast<linalg::BatchMatmulOp>(op)) {
+    inputType = batchMatmulOp.getInputOperand(0)->get().getType();
+
+    // First dimension is the batch dimension. We don't check the shape batch.
+    lhsShape =
+        getUntiledShape(batchMatmulOp.getInputOperand(0)->get()).drop_front(1);
+    rhsShape =
+        getUntiledShape(batchMatmulOp.getInputOperand(1)->get()).drop_front(1);
+
+    // First tile dimensions should be 1 for batched, use remaining dimensions
+    // for comparisons.
+    firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+    if (firstLevelTileSizes[0] != 1) {
+      op->emitError("Received first tile dimension of ")
+          << firstLevelTileSizes[0] << " instead of 1 for " << pipelineName;
+    }
+    firstLevelTileSizes = {firstLevelTileSizes[1], firstLevelTileSizes[2],
+                           firstLevelTileSizes[3]};
+  } else {
+    return success(); // Only verify batched and unbatched matmul.
+  }
```

Comment:
I think this logic could be simplified. It could be something like:
```
if(!isa<linalg::MatmulOp, linalg::BatchMatmulOp>(op)) {
  return success();
}
inputType = matmulOp.getInputOperand(0)->get().getType();
lhsShape = getUntiledShape(matmulOp.getInputOperand(0)->get());
rhsShape = getUntiledShape(matmulOp.getInputOperand(1)->get());
firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
if(linalg::BatchMatmulOp batchMatmulOp =
                 dyn_cast<linalg::BatchMatmulOp>(op)) {
// drop batch dim, and verify batch tile size
...
}
```